ignite/metrics/ssim.py

```diff
@@ -158,17 +159,31 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
             )
 
         channel = y_pred.size(1)
-        if len(self._kernel.shape) < 4:
-            self._kernel = self._kernel.expand(channel, 1, -1, -1).to(device=y_pred.device)
+
+        self._kernel = self._kernel.expand(channel, 1, -1, -1)
+        kernel = self._kernel
+
+        if y_pred.device != self._kernel.device:
+            warnings.warn(
+                "The metric's device should be the same than your update tensors. See SSIM() device argument.",
+                RuntimeWarning,
+            )
+
+            if self._kernel.device == torch.device("cpu"):
+                kernel = kernel.to(device=y_pred.device)
+
+            if y_pred.device == torch.device("cpu"):
+                y_pred = y_pred.to(device=self._kernel.device)
+                y = y.to(device=self._kernel.device)
 
         y_pred = F.pad(y_pred, [self.pad_w, self.pad_w, self.pad_h, self.pad_h], mode="reflect")
         y = F.pad(y, [self.pad_w, self.pad_w, self.pad_h, self.pad_h], mode="reflect")
 
-        if y_pred.dtype != self._kernel.dtype:
-            self._kernel = self._kernel.to(dtype=y_pred.dtype)
+        if y_pred.dtype != kernel.dtype:
+            kernel = kernel.to(dtype=y_pred.dtype)
```

Comment:
I was also thinking about dtype and was wondering if we could get predicted image as uint8 dtype ?

Reply:
that is absolutely right, scikit-image supports it.

Reply:
https://github.com/scikit-image/scikit-image/blob/v0.21.0/skimage/metrics/_structural_similarity.py#L219 images are converted to float from what I can see.

So maybe just putting
```python
if y_pred.dtype == torch.int8:
    y_pred = y_pred / 255

# float to float conversion
if y_pred.dtype != kernel.dtype:
            y_pred = y_pred.to(dtype=kernel.dtype)
```

Reply:
Yes, this makes sense from skimage side. Also it would be rare to get uint8 output from an NN.
Also conv2d probably wont work on uint8 input neither...